File: arrays_and_hashing/04_group_anagrams.py
def test_group_anagrams():
    assert (
        sort(group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
        == sort([["bat"], ["nat", "tan"], ["ate", "eat", "tea"]])
    )
    assert (
        sort(group_anagrams(["anagram", "nagaram", "margana"]))
        == sort([["anagram", "nagaram", "margana"]])
    )
    assert sort(group_anagrams([""])) == sort([[""]])
    assert sort(group_anagrams(["a"])) == sort([["a"]])
    assert sort(group_anagrams(["", ""])) == sort([["", ""]])
    assert sort(group_anagrams(["a", "a"])) == sort([["a", "a"]])
    assert sort(group_anagrams(["a", "b"])) == sort([["a"], ["b"]])
    assert sort(group_anagrams(["ab", "ba"])) == sort([["ab", "ba"]])


def sort(group):
    if len(group) < 2:
        return sorted(group)
    return sorted([sorted(words) for words in group])
    
    
# solution 1
from collections import defaultdict
def group_anagrams(strs: list[str]) -> list[list[str]]:
    grouped_anagrams = []
    for word in strs:
        chars = defaultdict(int)
        for char in word:
            chars[char] += 1
        for group in grouped_anagrams:
            if group[0] == chars:
                group[1].append(word)
                break
        else:
            grouped_anagrams.append([chars, [word]])
    groups = [group[1] for group in grouped_anagrams]
    return groups


# solution 2
from collections import defaultdict
def group_anagrams(strs: list[str]) -> list[list[str]]:
    grouped_anagrams = defaultdict(list)
    for word in strs:
        chars = defaultdict(int)
        for char in word:
            chars[char] += 1
        hashable_chars = tuple(sorted(chars.items()))
        grouped_anagrams[hashable_chars].append(word)
    return grouped_anagrams.values()

# solution 3
from collections import defaultdict
def group_anagrams(strs: list[str]) -> list[list[str]]:
    grouped_anagrams = defaultdict(list)
    for word in strs:
        chars = [0] * 26
        for char in word:
            chars[ord(char) - ord("a")] += 1
        grouped_anagrams[tuple(chars)].append(word)
    return grouped_anagrams.values()


if __name__ == "__main__":
    test_group_anagrams()

File: 01_arrays_and_hashing/04_group_anagrams.cpp
#include <cassert>
#include <cstdio>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>
#include <iterator>
#include <algorithm>

using std::vector, std::string;

string make_hashable(const vector<int>& vec) {
	std::ostringstream oss;
	std::copy(vec.begin(), vec.end(), std::ostream_iterator<int>(oss, " "));
	return oss.str();
}

vector<vector<string>> group_anagrams(vector<string>& strs) {
	std::unordered_map<string, vector<string>> grouped_anagrams;
	for (const string& word : strs) {
		vector<int> chars(26, 0);
		for (const char& c : word) {
			chars[c - 'a']++;
		}
		grouped_anagrams[make_hashable(chars)].push_back(word);
	}

	vector<vector<string>> result;
	for (const auto& [key, value] : grouped_anagrams) {
		result.push_back(value);
	}
	return result;
}

void test_make_hashable()
{
	vector<int> v = {2, 4, 6};
	assert(make_hashable(v) == "2 4 6 ");

	printf("All tests passed for make_hashable.\n");
}

        
void test_group_anagrams()
{
	vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
	vector<vector<string>> result = group_anagrams(strs);
	vector<vector<string>> expected = {{"bat"}, {"nat", "tan"}, {"ate", "eat", "tea"}};
	for (auto &v : result) {
		std::sort(v.begin(), v.end());
	}
	std::sort(result.begin(), result.end());
	for (auto &v : expected) {
		std::sort(v.begin(), v.end());
	}
	std::sort(expected.begin(), expected.end());

	assert(result == expected);

	strs = {""};
	expected = {{""}};
	assert(group_anagrams(strs) == expected);

	strs = {"a"};
	expected = {{"a"}};
	assert(group_anagrams(strs) == expected);

	strs = {"bdddddddddd","bbbbbbbbbbc"};
	expected = {{"bbbbbbbbbbc"},{"bdddddddddd"}};
	result = group_anagrams(strs);
	std::sort(result.begin(), result.end());
	assert(group_anagrams(strs) == expected);

	printf("All tests passed for group_anagrams.\n");
}

int main()
{
	test_make_hashable();
	test_group_anagrams();
	return 0;
}
